### PR TITLE
 Add caching to BaseChatModel (issue #1644)

### LINKED ISSUE #1644: Add caching support to BaseChatModel
Looks like `BaseLLM` supports caching (via `langchain.llm_cache`), but `BaseChatModel` does not.

### PR DESCRIPTION
**Pull request by UmerHA**
#  Add caching to BaseChatModel
Fixes #1644

(Sidenote: While testing, I noticed we have multiple implementations of Fake LLMs, used for testing. I consolidated them.)

## Who can review?
Community members can review the PR once tests pass. Tag maintainers/contributors who might be interested:
Models
- @hwchase17
- @agola11

Twitter: [@UmerHAdil](https://twitter.com/@UmerHAdil) | Discord: RicChilligerDude#7589

### CODE REVIEW

**Comment by abdulzain6** [comment]
Someone please take a look at this. Really need this :) Thanks

**Comment by realjustinwu** [comment]
Need this too.

**Comment by ielmansouri** [comment]
I hope it's reviewed soon, we need caching for ChatModels !

**Comment by Rienkim** [comment]
```python
langchain.llm_cache = SQLiteCache(database_path=".langchain.db")
chat = ChatOpenAI(temperature=0, openai_api_key=get_openai_api_key())
messages = [
    SystemMessage(content="You are a helpful assistant that translates English to French."),
    HumanMessage(content="I love programming.")
]
start = time.time()
print(chat(messages))
print(f"first time = {time.time() - start}")
start = time.time()
print(chat(messages))
print(f"first time = {time.time() - start}")
```

I test this code with this PR. First request miss cache, so It works. But, second request hit cache, and error occur.

```
cls = <class 'langchain.schema.ChatGeneration'>
values = {'generation_info': None, 'text': "J'adore la programmation."}

    @root_validator
    def set_text(cls, values: Dict[str, Any]) -> Dict[str, Any]:
>       values["text"] = values["message"].content
E       KeyError: 'message'
```

With InMemoryCache, test code work fine

```python
langchain.llm_cache = InMemoryCache()
chat = ChatOpenAI(temperature=0, openai_api_key=get_openai_api_key())
messages = [
    SystemMessage(content="You are a helpful assistant that translates English to French."),
    HumanMessage(content="I love programming.")
]
start = time.time()
print(chat(messages))
print(f"first time = {time.time() - start}")
start = time.time()
print(chat(messages))
print(f"first time = {time.time() - start}")
```

**Comment by UmerHA** [comment]
Hey @Rienkim, thanks for pointing that out! I'll take a look & add more tests that use more different caching options.

ETA should be this week. In the meanwhile, I'll turn this PR into a draft.

**Comment by UmerHA** [comment]
@Rienkim Fixed it & added more tests

**Comment by sam-cohan** [comment]
Hi folks, what is the latest on this? Is there a timeline for merging this?

**Comment by UmerHA** [comment]
> @UmerHA thank you for working on this.
> 
> I found that `_combine_llm_outputs` implemented in this PR can be an issue with `OpenAICallbackHandler` https://github.com/hwchase17/langchain/blob/master/langchain/callbacks/openai_info.py#LL99C42-L99C42

Can you post example code with actual & expected behavior? To me it seems fine, but maybe I'm missing something:

```
langchain.llm_cache = SQLiteCache(".langchain.test.db")
langchain.llm_cache.clear()

oai_cb = OpenAICallbackHandler()

llm = ChatOpenAI(callbacks=[oai_cb])

messages = [
    SystemMessage(content="You are a helpful assistant that translates English to French."),
    HumanMessage(content="I love programming.")
]

print(llm(messages))
print(oai_cb)

print(llm(messages))
print(oai_cb)

print(llm(messages))
print(oai_cb)
```

gives first
```
content="J'adore programmer." additional_kwargs={} example=False
Tokens Used: 34
	Prompt Tokens: 28
	Completion Tokens: 6
Successful Requests: 1
Total Cost (USD): $6.8e-05
```
then
```
content="J'adore programmer." additional_kwargs={} example=False
Tokens Used: 34
	Prompt Tokens: 28
	Completion Tokens: 6
Successful Requests: 2
Total Cost (USD): $6.8e-05
```
and then
```
content="J'adore programmer." additional_kwargs={} example=False
Tokens Used: 34
	Prompt Tokens: 28
	Completion Tokens: 6
Successful Requests: 3
Total Cost (USD): $6.8e-05
```

What would you expect to be different? :)

**Comment by pors** [comment]
> > @UmerHA thank you for working on this.
> > I found that `_combine_llm_outputs` implemented in this PR can be an issue with `OpenAICallbackHandler` https://github.com/hwchase17/langchain/blob/master/langchain/callbacks/openai_info.py#LL99C42-L99C42
> 
> Can you post example code with actual & expected behavior? To me it seems fine, but maybe I'm missing something:
> 
> ```
> langchain.llm_cache = SQLiteCache(".langchain.test.db")
> langchain.llm_cache.clear()
> 
> oai_cb = OpenAICallbackHandler()
> 
> llm = ChatOpenAI(callbacks=[oai_cb])
> 
> messages = [
>     SystemMessage(content="You are a helpful assistant that translates English to French."),
>     HumanMessage(content="I love programming.")
> ]
> 
> print(llm(messages))
> print(oai_cb)
> 
> print(llm(messages))
> print(oai_cb)
> 
> print(llm(messages))
> print(oai_cb)
> ```
> 
> gives first
> 
> ```
> content="J'adore programmer." additional_kwargs={} example=False
> Tokens Used: 34
> 	Prompt Tokens: 28
> 	Completion Tokens: 6
> Successful Requests: 1
> Total Cost (USD): $6.8e-05
> ```
> 
> then
> 
> ```
> content="J'adore programmer." additional_kwargs={} example=False
> Tokens Used: 34
> 	Prompt Tokens: 28
> 	Completion Tokens: 6
> Successful Requests: 2
> Total Cost (USD): $6.8e-05
> ```
> 
> and then
> 
> ```
> content="J'adore programmer." additional_kwargs={} example=False
> Tokens Used: 34
> 	Prompt Tokens: 28
> 	Completion Tokens: 6
> Successful Requests: 3
> Total Cost (USD): $6.8e-05
> ```
> 
> What would you expect to be different? :)

@deepblue ^^^ (in case you missed it)

**Comment by jakobsa** [comment]
@deepblue could you elaborate on your concerns? I am waiting for the feature :)

**Comment by deepblue** [comment]
Just realized I made an error in my code- used `BaseChatModel._combine_llm_outputs` in this PR instead of the default `ChatOpenAI._combine_llm_outputs` for testing. Retested with the right method and it's all clear. The original method doesn't affect our subclassed code. Apologies for the mix-up and the late response.

I tested and confirmed that it's working as expected

**Comment by pors** [comment]
@hwchase17
@agola11

Good to go?

**Comment by kaikun213** [comment]
@hwchase17 
@agola11 

Any update on this?

File: langchain/chat_models/base.py [hunk 1]
```diff
@@ -59,7 +110,24 @@ class Config:
         arbitrary_types_allowed = True
 
     def _combine_llm_outputs(self, llm_outputs: List[Optional[dict]]) -> dict:
-        return {}
+        """Combine general llm outputs by aggregating them into lists
```
Comment (hwchase17) [inline langchain/chat_models/base.py:113]: this seems separate? am going to revert this

File: langchain/chat_models/base.py [hunk 2]
```diff
@@ -178,6 +184,96 @@ async def agenerate_prompt(
...
+                ]
+                return ChatResult(generations=cache_val_for_chat)
+            else:
+                if new_arg_supported:
+                    result = self._generate(
+                        messages, stop=stop, run_manager=run_manager, **kwargs
+                    )
+                else:
+                    result = self._generate(messages, stop=stop, **kwargs)
+                langchain.llm_cache.update(prompt, llm_string, result.generations)
```
Comment (dev2049) [inline langchain/chat_models/base.py:229]: won't we be storing ChatGenerations if _generate creates ChatGenerations? in which case do we need to do extra parsing in 218

File: docs/snippets/modules/model_io/models/llms/how_to/llm_caching.mdx [hunk 3]
```diff
@@ -14,7 +14,7 @@ from langchain.cache import InMemoryCache
 langchain.llm_cache = InMemoryCache()
 
 # The first time, it is not yet in cache, so it should take longer
-llm("Tell me a joke")
+llm.predict("Tell me a joke")
```
Comment (nfcampos) [inline docs/snippets/modules/model_io/models/llms/how_to/llm_caching.mdx:17]: Why is this changing?

File: langchain/cache.py [hunk 4]
```diff
@@ -163,7 +179,7 @@ def update(self, prompt: str, llm_string: str, return_val: RETURN_VAL_TYPE) -> N
     def clear(self, **kwargs: Any) -> None:
         """Clear cache."""
         with Session(self.engine) as session:
-            session.execute(self.cache_schema.delete())
+            session.query(self.cache_schema).delete()
```
Comment (nfcampos) [inline langchain/cache.py:182]: Why is this changing ?

Comment (hwchase17) [reply langchain/cache.py:182]: yeah not sure, from previous pr, ill revert

File: docs/snippets/modules/model_io/models/llms/how_to/llm_caching.mdx (hunk 3 above)
Comment (hwchase17) [reply docs/snippets/modules/model_io/models/llms/how_to/llm_caching.mdx:17]: to use the consist predict/predict_messages interface

**Comment by UmerHA** [comment]
> Getting the below error when I use `MomentoCache`. @UmerHA please let me know if this is a bug or if I am doing anything wrong.
> 
> Edit: The error pops up for all calls after the cache has atleast 1 key set.
> 
> ```shell
> @root_validator
>     def set_text(cls, values: Dict[str, Any]) -> Dict[str, Any]:
> >       values["text"] = values["message"].content
> E       KeyError: 'message'
> ```
> 
> Code:
> 
> ```python
> import langchain
> from datetime import timedelta
> from langchain.cache import MomentoCache
> 
> langchain.llm_cache = MomentoCache.from_client_params("langchain_momento", imedelta(days=1))
> 
> # Further code for constructing and calling the chain using ChatOpenAI
> ```

Can you post the full code, error message, and stack trace?

**Pull request merged**